src/jupyter/contents/sessions.unit.test.ts

```diff
@@ -0,0 +1,449 @@
+      // Cast needed due to overload.
+      (assignmentManager.getServers as sinon.SinonStub)
+        .withArgs('extension')
+        .returns(deferred.promise);
+
+      const firstPromise = manager.getOrCreate(DEFAULT_SERVER.endpoint);
+      const secondPromise = manager.getOrCreate(DEFAULT_SERVER.endpoint);
+      deferred.resolve([DEFAULT_SERVER]);
+
+      const firstRes = await firstPromise;
+      const secondRes = await secondPromise;
+      expect(firstRes).to.equal(secondRes);
+    });
+
+    it('returns established connections', async () => {
+      toggleAuth(AuthState.SIGNED_IN);
+      // Cast needed due to overload.
+      (assignmentManager.getServers as sinon.SinonStub).resolves([
+        DEFAULT_SERVER,
+      ]);
+      const first = await manager.getOrCreate(DEFAULT_SERVER.endpoint);
+
+      const second = await manager.get(DEFAULT_SERVER.endpoint);
+
+      expect(first).to.equal(second);
+    });
+
+    it('can establish a new connection if a previous attempt failed', async () => {
+      toggleAuth(AuthState.SIGNED_IN);
+      // Cast needed due to overload.
+      (assignmentManager.getServers as sinon.SinonStub)
+        .withArgs('extension')
+        .onFirstCall()
+        .rejects()
+        .onSecondCall()
+        .resolves([DEFAULT_SERVER]);
+
+      await expect(manager.getOrCreate(DEFAULT_SERVER.endpoint)).to.eventually
+        .be.rejected;
+      await expect(manager.getOrCreate(DEFAULT_SERVER.endpoint)).to.eventually
+        .be.fulfilled;
+    });
+  });
+
+  describe('drop', () => {
+    let listener: sinon.SinonStub<[string]>;
+
+    function waitForClientDisposed() {
+      return new Promise<void>((r) => {
+        jupyterClientStub.dispose.callsFake(() => {
+          r();
+        });
+      });
+    }
+
+    beforeEach(() => {
+      toggleAuth(AuthState.SIGNED_IN);
+      listener = sinon.stub();
+      manager.onDidRevokeConnection(listener);
+    });
+
+    it('throws if disposed', () => {
+      manager.dispose();
+
+      expect(() => manager.drop('foo')).to.throw(/disposed/);
+    });
+
+    it("returns false and no-ops when there's no matching connection", () => {
+      expect(manager.drop('foo')).to.be.false;
+    });
+
+    it('drops a managed connection', async () => {
+      // Cast needed due to overload.
+      (assignmentManager.getServers as sinon.SinonStub).resolves([
+        DEFAULT_SERVER,
+      ]);
+      await manager.getOrCreate(DEFAULT_SERVER.endpoint);
+      const clientDisposed = waitForClientDisposed();
+
+      const dropped = manager.drop(DEFAULT_SERVER.endpoint);
+
+      expect(dropped).to.be.true;
+      await expect(clientDisposed).to.be.eventually.fulfilled;
+      sinon.assert.calledOnceWithExactly(listener, DEFAULT_SERVER.endpoint);
+    });
+
+    it('drops a managed connection silently', async () => {
+      // Cast needed due to overload.
+      (assignmentManager.getServers as sinon.SinonStub).resolves([
```

Comment:
*nit* maybe declare this as a stub in the beforeEach? Could slightly de-dupe

Reply:
Done!